### ISSUE TITLE
Question about a deviation from the standard in adoption agency algorithm

### ISSUE THREAD

**Issue by stasoid** [lexbor/lexbor#187]
I compared your implementation of adoption agency algorithm with the standard and the only deviation I found is [`continue` at State 14.5]. This corresponds to step 4.13.4 in [the current version of the standard], which doesn't have the `continue`.

I believe this is an error in the standard because commenting out that line fails one of the html5lib tests: `<div><a><b><u><i><code><div></a>` (test 15 from adoption01.ton).

I noticed that this error [existed] in the standard 1 year before first commit to myhtml.
Also, this line has a [spec comment in myhtml] which doesn't mention `continue` either.
It looks like you fixed the test despite what standard says.
Am I right? I want to be sure before creating an issue at whatwg.

[`continue` at State 14.5]: https://github.com/lexbor/lexbor/blob/8f889f68886/source/lexbor/html/tree.c#L1535
[the current version of the standard]: https://html.spec.whatwg.org/multipage/parsing.html#adoption-agency-algorithm
[existed]: https://www.w3.org/TR/2014/REC-html5-20141028/syntax.html#adoption-agency-algorithm
[spec comment in myhtml]: https://github.com/lexborisov/myhtml/blob/4f98bb1562c/source/myhtml/tree.c#L1799..L1800

**Comment by lexborisov** [comment]
Hi @stasoid 

This all relates to the specifics of reading and understanding the spec.

Let's take two paragraphs:

> 4.13.3: If inner loop counter is greater than 3 and node is in the [list of active formatting elements](https://html.spec.whatwg.org/multipage/parsing.html#list-of-active-formatting-elements), then remove node from the [list of active formatting elements](https://html.spec.whatwg.org/multipage/parsing.html#list-of-active-formatting-elements).

But in the next paragraph it says:

> 4.13.4: If node is not in the [list of active formatting elements](https://html.spec.whatwg.org/multipage/parsing.html#list-of-active-formatting-elements), then remove node from the [stack of open elements](https://html.spec.whatwg.org/multipage/parsing.html#stack-of-open-elements) and [continue](https://infra.spec.whatwg.org/#iteration-continue).

That is, in paragraph 4.13.3 we remove "active formatting elements", and in 4.13.4 we do a check for existence in "active formatting elements".

But, since we had the node "active formatting element", we don't need to do the check in 4.13.4 and we don't need to remove it from the "stack of open elements" because the node is "active formatting element" and it won't be in the "stack of open elements".

So we just make a continue in 4.13.3.
In other words, it's such an optimization.

**Comment by stasoid** [comment]
I think "active formatting element" can be in the "stack of open elements", I don't know why you assume it can't. So we do need to remove it from the "stack of open elements". Also, as I mentioned, after removing `continue` at 4.13.4 tests start failing - it changes the logic.

**Comment by lexborisov** [comment]
@stasoid 

> I think "active formatting element" can be in the "stack of open elements", I don't know why you assume it can't.

Show me a case in the spec where this is possible? I haven't found such cases, but I could be wrong.

> Also, as I mentioned, after removing continue at 4.13.4 tests start failing - it changes the logic. (Btw step numbers in your message are a little wrong)

You are wrong, all tests are just as successful.

I changed the code and tested it. Nothing changed.

```C
            /* State 14.5 */
            if (inner_loop_counter > 3 && is) {
                lxb_html_tree_active_formatting_remove_by_node(tree, node);

                is = false;
            }

            /* State 14.6 */
            if (is == false) {
                lxb_html_tree_open_elements_remove_by_node(tree, node);

                continue;
            }
```

Maybe I missed something?

**Comment by stasoid** [comment]
You are absolutely right, sorry for wasting your time. I didn't do the `is = false`.

I thought that "active formatting element" can be in the "stack of open elements" because of step 4.4 of adoption agency algorithm: there is a parse error if "formatting element" is *not* in the "stack of open elements". But apparently, the situation in 4.13.4 is different, I don't know why.

Thank you for your answers.
